src/Mobs/Monster.cpp

```diff
@@ -122,101 +120,124 @@ void cMonster::SpawnOn(cClientHandle & a_Client)
+				return false;
 			}
-			if (m_Path->IsLastPoint())
+			else if (!m_Path->IsLastPoint() && (m_Path->IsFirstPoint() || ReachedDestination()))  // Have we arrived at the next cell, as denoted by m_Destination?
 			{
-				ResetPathFinding();
+				m_Destination = Vector3d(0.5, 0, 0.5) + m_Path->GetNextPoint();
+				m_GiveUpCounter = 40;  // Give up after 40 ticks (2 seconds) if failed to reach m_Destination.
 			}
-			break;
-
+			return true;
 		}
 	}
-}
-
 
-
-
-
-/* Currently, the mob will only start moving to a new position after the position it is
-currently going to is reached. */
-void cMonster::MoveToPosition(const Vector3d & a_Position)
-{
-	m_FinalDestination = a_Position;
-	m_bMovingToDestination = true;
+	return false;
 }
 
 
 
 
 
-void cMonster::StopMovingToPosition()
+void cMonster::MoveToWayPoint(cChunk & a_Chunk)
 {
-	m_bMovingToDestination = false;
-	ResetPathFinding();
-}
-
-
-
+	if (m_JumpCoolDown == 0)
+	{
+		// We're not moving and waypoint is above us, it means we are hitting something and we should jump.
+		if ((GetSpeedX() < 0.1) && (GetSpeedY() < 0.1) && DoesPosYRequireJump(FloorC(m_Destination.y)))
+		{
+			if (IsOnGround() || IsSwimming())
+			{
+				m_bOnGround = false;
+				m_JumpCoolDown = 20;
+				// TODO: Change to AddSpeedY once collision detection is fixed - currently, mobs will go into blocks attempting to jump without a teleport
+				AddPosY(1.6);  // Jump!!
+				SetSpeedX(3.2 * (m_Destination.x - GetPosition().x));  // Move forward in a preset speed.
+				SetSpeedZ(3.2 * (m_Destination.z - GetPosition().z));  // The numbers were picked based on trial and error and 1.6 and 3.2 are perfect.
+			}
+		}
+	}
+	else
+	{
+		--m_JumpCoolDown;
+	}
 
+	Vector3d Distance = m_Destination - GetPosition();
+	if ((Distance.x != 0) || (Distance.z != 0))
+	{
+		Distance.y = 0;
+		Distance.Normalize();
 
-bool cMonster::IsCoordinateInTraversedList(Vector3i a_Coords)
-{
-	return (std::find(m_TraversedCoordinates.begin(), m_TraversedCoordinates.end(), a_Coords) != m_TraversedCoordinates.end());
+		if (m_bOnGround)
+		{
+			Distance *= 2.5f;
+		}
+		else if (IsSwimming())
+		{
+			Distance *= 1.3f;
+		}
+		else
+		{
+			// Don't let the mob move too much if he's falling.
+			Distance *= 0.25f;
+		}
+		// Apply walk speed:
+		Distance *= m_RelativeWalkSpeed;
+		/* Reduced default speed.
+		Close to Vanilla, easier for mobs to follow m_Destinations, hence
+		better pathfinding. */
+		Distance *= 0.5;
+		AddSpeedX(Distance.x);
+		AddSpeedZ(Distance.z);
+	}
 }
 
 
 
 
 
-/* No one should call this except the pathfinder orthe monster tick or StopMovingToPosition.
-Resets the pathfinder, usually starting a brand new path, unless called from StopMovingToPosition. */
-void cMonster::ResetPathFinding(void)
+void cMonster::MoveToPosition(const Vector3d & a_Position)
 {
-	if (m_Path != nullptr)
+	if ((m_FinalDestination - a_Position).Length() > 0.25)
 	{
-		delete m_Path;
-		m_Path = nullptr;
+		ResetPathFinding();
```

Comment:
But not having a recalculation will also mean that if a player does move to a place where a different path is needed, for example, up a ladder where mobs can't follow, and there is no easy line-of-sight route, the mob will first go to the old position, realise it's wrong, and potentially have to go a long way back to a correct route if it's a complex maze and such.

Meanwhile, the player can simply move back down the ladder, and the mob will not realise, since its pathfinding isn't being reset. It will instead head down the route to get to where the player was when they headed up the ladder.

Perhaps a way to fix the 'mob stands there while path is calculated' problem is to have two Path variables. When a mob spots a player, it remains in the IDLE state, and tasks one cPath to pathfind to the player, whilst the other handles normal idling behaviour; random walking and such. If the player could be navigated to, the state is set to ATTACKING and the mob moves along the calculated route. If no route was found, the mob sets a variable indicating that the current target cannot be reached (continuing with idling behaviour), which is not cleared until the target moves position.

What would you suggest?